src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+    // against that changing in the future.
+    if (!fnHandle->IsFunction() || (!privHandle.IsEmpty() && privHandle->IsPrivate())) {
+      // We can't return this because it's not a function, or if it is, it's private.
+      return kj::str("The RPC receiver does not implement the requested method."_kj);
+    }
+    return fnHandle.As<v8::Function>();
+  }
+
+  // Deserializes the arguments and passes them to the given function.
+  v8::Local<v8::Value> invokeFn(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, kj::heapArray(serializedArgs)).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      auto arguments = kj::heapArray<v8::Local<v8::Value>>(args.size());
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
+    }
+  };
+
+  // The serialized JS value and a bool to discern whether we've thrown an exception.
+  // Serializing may throw if the data we are serializing is not structured-cloneable.
+  // In this case, we will just serialize the exception message and set `caughtException`.
+  struct SerializedResult {
+    kj::Array<kj::byte> data;
+    bool caughtException;
+  };
+
+  // Serializes the return value of our method.
+  SerializedResult serialize(jsg::Lock& js, jsg::JsValue value) {
+    kj::Maybe<SerializedResult> result;
+
+    js.tryCatch([&] {
+      auto data = serializeV8(js, value.addRef(js).getHandle(js));
+      // Serialized okay!
+      result = SerializedResult { .data = kj::mv(data), .caughtException = false };
+    }, [this, &js, &result](jsg::Value exception) {
+      auto fullError = kj::str(jsg::check(exception.getHandle(js)->ToString(js.v8Context())));
+
+      // We may need to slice part of the message so we don't have `Error: Error: msg`.
+      kj::Maybe<kj::StringPtr> errorMessage(fullError);
+      if (fullError.startsWith("Error: "_kj)) {
+        errorMessage = fullError.slice(kj::size("Error: "_kj));
+      }
+
+      // Failed to serialize data.
+      auto failedResult = serialize(js, js.error(KJ_REQUIRE_NONNULL(errorMessage)));
+      // We are in the error branch, so we must set caughtException.
+      failedResult.caughtException = true;
+      result = kj::mv(failedResult);
+    });
+    return kj::mv(KJ_REQUIRE_NONNULL(result));
+  }
+
+  // Invokes the given function and serializes the result for RPC.
+  kj::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
```

Comment:
Again here, you are passing a `jsg::Lock` into a coroutine... this causes problems. It happens not to be an issue here because you don't happen to invoke any V8 APIs after the first `co_await`, but it's quite fragile.

I think this function could actually be returning `jsg::Promise` instead of `kj::Promise`. Unfortunately we do not support coroutines for `jsg::Promise` yet, but at least you aren't mixing different event loops that way.

Reply:
https://github.com/cloudflare/workerd/pull/1028#discussion_r1412626362